solr/prometheus-exporter/src/java/org/apache/solr/prometheus/collector/MetricSamples.java

```diff
@@ -19,24 +19,42 @@
 
 import io.prometheus.client.Collector;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.stream.Collectors;
 
 public class MetricSamples {
 
   private final Map<String, Collector.MetricFamilySamples> samplesByMetricName;
 
+  private final Set<String> sampleMetricsCache;
+
   public MetricSamples(Map<String, Collector.MetricFamilySamples> input) {
     samplesByMetricName = input;
+    sampleMetricsCache = new HashSet<>();
+    for (Collector.MetricFamilySamples metricFamilySamples : input.values()) {
+      addSamplesToCache(metricFamilySamples);
+    }
   }
 
   public MetricSamples() {
     this(new HashMap<>());
   }
 
-  public void addSamplesIfNotPresent(String metricName, Collector.MetricFamilySamples samples) {
-    samplesByMetricName.putIfAbsent(metricName, samples);
+  private void addSamplesToCache(Collector.MetricFamilySamples metricFamilySamples) {
+    for (Collector.MetricFamilySamples.Sample sample : metricFamilySamples.samples) {
+      sampleMetricsCache.add(sample.toString());
+    }
+  }
+
+  public void addSamplesIfNotPresent(
+      String metricName, Collector.MetricFamilySamples metricFamilySamples) {
+    if (!samplesByMetricName.containsKey(metricName)) {
+      samplesByMetricName.put(metricName, metricFamilySamples);
```

Comment:
There used to be a putIfAbsent call here, which is more elegant (and efficient) than a check and put.  You can still use that; just rely on the return value of putIfAbsent being null to check if you then should call addSamplesToCache.  If you think it's still unclear, you could add a comment.